src/control/server/storage/scm/provider.go

```diff
@@ -0,0 +1,564 @@
+		return nil, errors.Wrapf(err, "failed to format %s", src)
+	}
+
+	res, err := p.MountDcpm(src, req.Mountpoint)
+	if err != nil {
+		return nil, err
+	}
+	return &FormatResponse{
+		Mountpoint: res.Target,
+		Formatted:  res.Mounted,
+	}, nil
+}
+
+func (p *Provider) MountDcpm(device, target string) (*MountResponse, error) {
+	return p.mount(device, target, dcpmFsType, defaultMountFlags, dcpmMountOpts)
+}
+
+func (p *Provider) MountRamdisk(target string, size uint) (*MountResponse, error) {
+	var opts string
+	if size > 0 {
+		opts = fmt.Sprintf("size=%dg", size)
+	}
+
+	return p.mount(ramFsType, target, ramFsType, defaultMountFlags, opts)
+}
+
+func (p *Provider) Mount(req MountRequest) (*MountResponse, error) {
+	return p.mount(req.Source, req.Target, req.FsType, req.Flags, req.Data)
+}
+
+func (p *Provider) mount(src, target, fsType string, flags uintptr, opts string) (*MountResponse, error) {
+	if err := os.Mkdir(target, defaultMountPointPerms); err != nil {
+		return nil, errors.Wrapf(err, "failed to create mountpoint %s", target)
+	}
+
+	p.log.Debugf("mount %s->%s (%s) (%s)", src, target, fsType, opts)
+	if err := p.sys.Mount(src, target, fsType, flags, opts); err != nil {
+		return nil, errors.Wrapf(err, "mount %s->%s failed", src, target)
+	}
+
+	return &MountResponse{
+		Target:  target,
+		Mounted: true,
+	}, nil
+}
+
+func (p *Provider) Unmount(req MountRequest) (*MountResponse, error) {
+	return p.unmount(req.Target, int(req.Flags))
+}
+
+func (p *Provider) unmount(target string, flags int) (*MountResponse, error) {
+	if err := p.sys.Unmount(target, flags); err != nil {
+		return nil, errors.Wrapf(err, "failed to unmount %s", target)
+	}
+
+	return &MountResponse{
+		Target:  target,
+		Mounted: false,
+	}, nil
+}
+
+func (p *Provider) IsMounted(target string) (bool, error) {
+	return p.sys.IsMounted(target)
+}
+
+func (p *Provider) Format(req FormatRequest) (*FormatResponse, error) {
+	if req.Ramdisk != nil && req.Dcpm != nil {
+		return nil, FaultFormatConflictingParam
+	}
+	if req.Mountpoint == "" {
+		return nil, FaultFormatMissingMountpoint
+	}
+
+	check, err := p.CheckFormat(req)
+	if err != nil {
+		return nil, err
+	}
+	if check.Formatted {
+		if !req.Reformat {
+			return nil, FaultFormatNoReformat
+		}
+	}
+
+	if err := p.clearMount(req); err != nil {
+		return nil, errors.Wrap(err, "failed to clear existing mount")
+	}
+
+	switch {
+	case req.Ramdisk != nil:
+		return p.formatRamdisk(req)
+	case req.Dcpm != nil:
+		return p.formatDcpm(req)
+	default:
+		return nil, FaultFormatMissingParam
+	}
+}
+
+func (p *Provider) Update(req UpdateRequest) (*UpdateResponse, error) {
+	return nil, nil
+}
+
+func (p *Provider) createScanResponse() *ScanResponse {
+	p.RLock()
+	defer p.RUnlock()
+
+	return &ScanResponse{
+		State:      p.lastState,
+		Modules:    p.modules,
+		Namespaces: p.namespaces,
+	}
+}
+
+func (p *Provider) Scan(req ScanRequest) (*ScanResponse, error) {
```

Comment:
Would be nice to have documentation comments on all these public methods

Reply:
Yep, will add those before re-requesting.

Reply:
Done.